app/helpers/admin/stripe_helper.rb

```diff
@@ -0,0 +1,75 @@
+# require File.join(Rails.root, '/lib/oauth2/strategy/deauthorize')
+# require File.join(Rails.root, '/lib/oauth2/client')
+# require 'oauth2'
+module Admin
+  module StripeHelper
+
+    class << self
+      attr_accessor :client, :options
+    end
+
+    @options = {
+      :site => 'https://connect.stripe.com',
+      :authorize_url => '/oauth/authorize',
+      :deauthorize_url => '/oauth/deauthorize',
+      :token_url => '/oauth/token'
+    }
+
+    @client = OAuth2::Client.new(
+      ENV['STRIPE_CLIENT_ID'],
+      ENV['STRIPE_INSTANCE_SECRET_KEY'],
+      options
+    )
+
+    def get_stripe_token(code, options={})
+      StripeHelper.client.auth_code.get_token(code, options)
+    end
+
+    def authorize_stripe(enterprise_id, options={})
+      options = options.merge({enterprise_id: enterprise_id})
+      jwt = jwt_encode options
+      # State param will be passed back after auth
+      StripeHelper.client.auth_code.authorize_url(state: jwt, scope: 'read_write')
+    end
+
+    def deauthorize_stripe(account_id)
+      stripe_account = StripeAccount.find(account_id)
+      if stripe_account
+        # If the account is only connected to one Enterprise, make a request to remove it on the Stripe side
+        if StripeAccount.where(stripe_user_id: stripe_account.stripe_user_id).size == 1
+          response = deauthorize_request_for_stripe_id(stripe_account.stripe_user_id)
+          if response # Response from OAuth2 only returned if successful
+            stripe_account.destroy
+          end
+        else
+          stripe_account.destroy
+        end
+      end
+    end
+
+    def fetch_event_from_stripe(request)
+      event_json = JSON.parse(request.body.read)
+      # If the application has been deauthorised, we are no longer authorised to retrieve events for that account
+      # Left here in case it's useful for other webhooks
+      unless event_json["type"] == "account.application.deauthorized"
+        acct_param = event_json["user_id"] ? {"Stripe-Account" => event_json["user_id"]} : nil
+        Stripe::Event.retrieve(event_json["id"],acct_param)
+      else
+        Stripe::Event.construct_from(event_json)
+      end
+    end
+
+    def deauthorize_request_for_stripe_id(id)
+      StripeHelper.client.deauthorize(id).deauthorize_request
+    end
+
+    private
+    def jwt_encode payload
+      JWT.encode(payload, Openfoodnetwork::Application.config.secret_token, 'HS256')
+    end
+
+    def jwt_decode token
+      JWT.decode(token, Openfoodnetwork::Application.config.secret_token, true, algorithm: 'HS256')[0] # only returns the original payload
+    end
+  end
+end
```

Comment:
I stronlgy believe this module should actually be the abstraction the controllers are lacking. For my experience, I believe that payment gateways need to be encapsulated so as not to leak details throughout the whole app. I see that these methods could be independent service objects like `Stripe::Authorize`, `Stripe::Deauthorize`, etc.

Reply:
Resolved